## XP Points Challenge 2
CS 133  
Spring 2025  
v1.0

A message has been encoded in the dataset. Your taks is to complete the operations on the dataset to discover the hidden message.

1. Load the datset `xp-2.csv` into a dataframe and inspect it.

In [85]:
import pandas as pd
df = pd.read_csv("xp-2.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  1000 non-null   int64 
 1   id          1000 non-null   object
 2   username    1000 non-null   object
 3   gender      1000 non-null   object
 4   age         1000 non-null   int64 
 5   country     1000 non-null   object
 6   ip          1000 non-null   object
 7   online      1000 non-null   int64 
 8   model       1000 non-null   object
dtypes: int64(3), object(6)
memory usage: 70.4+ KB


There are some features of Pandas that we haven't gone over that could be useful for this problem and you may want to look them up:
1. The `DataFrame.where()` method
2. The `sum()` method
3. The `apply()` method
4. Converting a Numpy array into a string using `.str.cat()`

2. Create a new feature named `model` based on the age of the individual and the number of minutes spent online. If the individual is older than 50 and spends more than 100 minutes online assign them the value of `boomer`, otherwise they should be classified as `gamer`.
   

In [71]:
#Im certain there is a one line way to do this but honestly I find this more readable would love your input
def BorG(dataframe):
    if dataframe["age"] >50 and dataframe["online"] > 100:
        return "boomer"
    else:
        return "gamer"


df['model'] = df.apply(BorG, axis = 1)
df

Unnamed: 0                          id     username gender  age  \
0             0  01JMTV0RD07H2C0SYJEHCF0EEG       Lotlux      M   40   
1             1  01JMTV0RD1KT7R9ED50SPYYJ9F  Solarbreeze      F   96   
2             2  01JMTV0RD22T0A497N6JW1WA2P      Flexidy      M   80   
3             3  01JMTV0RD2J6NZQW6QZG5TZA4A   Voltsillam      M  108   
4             4  01JMTV0RD3P25R4YKK8AGE8WJM       Vagram      F   90   
..          ...                         ...          ...    ...  ...   
995         995  01JMTV0S31PKMA5NAGMJ0M37A5   Stronghold      M   84   
996         996  01JMTV0S32P8J391FHCWZMJFNF      Prodder      F   47   
997         997  01JMTV0S32QVX88ADFF42ECK2Q      Sonsing      M   72   
998         998  01JMTV0S3377B457TBTWSVB8GH     Bytecard      F   88   
999         999  01JMTV0S34YSJXJ7B34RQKG6NH      Prodder      F   15   

         country               ip  online   model  
0          China   246.220.74.249     188   gamer  
1           Oman    206.39.45.156     893  boomer  
2          Qatar    252.181.153.6     979  boomer  
3       Portugal  210.230.146.108     199  boomer  
4      Argentina     94.160.4.110     812  boomer  
..           ...              ...     ...     ...  
995    Indonesia    251.29.79.115     875  boomer  
996  Philippines   236.88.247.159     456   gamer  
997        China     218.76.5.199     334  boomer  
998  Afghanistan   210.107.44.142     433  boomer  
999   Bangladesh     169.41.21.53     488   gamer  

[1000 rows x 9 columns]

In [79]:
#Verifying that values are assigned properly otherwise the next part of the challenge wont work. I feel like below is close to the one line method I could have used in question 2.

check = df.query( '(age > 50) and (online > 100) and (model != "boomer") or (age <= 50) and (online <= 100) and (model != "gamer")' )
check #should be an empty dataframe

Empty DataFrame
Columns: [Unnamed: 0, id, username, gender, age, country, ip, online, model]
Index: []

3. To receive credit for this assignment you must create a single statement that performs all of the operations described below to uncover the hidden message. You may want to try doing the operations separately until you solve it but you must figure out a way to do it as a single statement.

   1. Group the dataframe by  `model` then `gender` then `country`
   2. Find the sum of the `online` column in each group
   3. Find the modulus of the sum and 123 in each group
   4. For each value use the Python `chr()` function to convert it to a printable character
   5. Convert the first 71 characters into a string and display it

In [83]:
#A, pass a list of columns. B say what column we want and sum the values. C is poorly worded but I think we take our number % 123. D cast as a character. E List comprehension vodoo

#A, pass a list of columns:
#df.groupby(['model', 'gender', 'country'])

#B With our new groupby dataframe tell the frame we want the online column and we want to add them, all:
#df.groupby(['model', 'gender', 'country'])["online"].sum()

#C Easy slap % 123 at the end:
#df.groupby(['model', 'gender', 'country'])["online"].sum() % 123

#D cast as a character? but we will need each piece of iteration so we cant do it until later. Below is wrong but my thought process:
#for i in range (71): chr(df.groupby(['model', 'gender', 'country'])["online"].sum() % 123)

#E Actually do the hard part, comprehend the list. Have to iterate before casting... then take each iteration into a string, .join() time:
''.join( chr(x) for x in (df.groupby(['model', 'gender', 'country'])["online"].sum() % 123))
#Customs be dammed im importing pandas into my code, who else is going to take build my dataframes??

"Why did customs arrest the data scientist? For trying to import pandas!C0D,wpt\x195C\tFMeYv5\x11Ahc-*D](-`UkmC',\x1fG9d&a#qXc\x18v5'Drv^#\x1e[s\x07yJ? &N*`_jNr\x11g\x02bEG\x05wlo\rwz\x1b:PDOv\x055\x1f\x03&L>y\nq\x1cSNR/;dtC\x0b4s\x07\x18\x0e]mVW\x1d5_[D\x1ahg\x0cGZh,+6O\\c\x19I\x03\x08dk!=(5\x1f\x14\x1fC7J\x04z5cS\x08\\\x01=8-viaSC6ytF%T\x07\x05CcBy*s0R,)YVC\x03C(\x05^L\x1dbz"

Since I got feedback about the given dataset I did this to clean the data. They are not part of the solution and have been commented out for an easier time hitting restart kernal and run of all cells.

In [84]:
#df.drop('model', axis = 1, inplace = True)
#df.info()

## Additonal semi-related comments:

I liked this challenge, but it took me far longer than the intended 30 minutes youd expect of the class, of course I am just one student so it could just be me who took more time. I wanted to give you props for the effort in getting the class involved with some active learning skill the lemur caught me off guard. I am curious to see your approach to those optional chapters we talked about, im still thinking the extra credit should be low (like 1 point for reading the chapter, 1 for the lab of that chapter, and 1 for the optional exam question for that chapter) my Physics class does 3% extra credit overall. However you decide to get (or not get) that extra credit is up to the student but 3% is fair. 

Feedback for this code:
Is this too much commenting? Not enough? Is there a better way to do this?

### Some other ideas for you that the class might hate but might be a good learning activity:

Group coding assignment: Similar to the Labs in format but I am thinking more of a take home idea. There should be some consideration as to preventing one person in the group from doing all the work. My remedy would be to require the use of a github repo as I think it would be a practical skill for people to get acustomed to and the barrier to entry is fairly low when using the desktop client (I would not expect anyone to be using git CLI). You could see who pushed commits and who didnt to see if the group work was "fair" At the same time this could be opening a can of worms trying to help folks understand github and diagnose merging pull requests this idea could get messy very quickly. This also could be a huge grading pain for you so maybe not the best idea.

Code Reviews and Unit tests:
Personally I would greatly appreciate getting into the habit, being corrected on, and shown what is considered best practices, I recall iloc vs loc or setting your random_state. We mostly already do this but I could see us posting our code somewhere for review and then breaking it apart in class (anonymized) to say what is good or what is bad about said code. Of course this is subjective but you've seen the errors of your 334 class what do you want us to get right before we get to that level? Before we head into 233 (I am hopeful for a fall session that way I could do 133, 233, and 334 back to back to back)

Practical Projects:
Again im pushing for projects but this might be over the course of an entire unit instead of the homeworks which cover a homework portion (or a whole unit like unit 1). It could be valuable for us to do every step of the data science lifecycle we saw in chapter one (minus data collection and likely very simplified) maybe thats saved for 233, either way would be neat.

With these suggestions my thought process is that 133 prepares the non CS for practical data science in their fields so that even if they never take another data science course they might still be able to do the core tasks of data science or at the bare minimum collect cleaner data. With that I would think that 233 is more geared towards the CS crowd or the people in this class that say "you know this is kinda useful in my field", and 334 is all in for CS folks. My intention for these suggestions isn't to add more work for the sake of adding more work but to be able to walk away from this class thinking that it was worth taking. I am huge proponent of learning by doing through practical experinces/assignments in the classroom that way if a student doesn't go on to 233 they can still take what they learned (even if they forget 90%) into their related science field with the ideas of where their data might go. Additionally I think having a scientest that at least considers where their data goes is far better than a scientest who only knows their area of study.

Just some thoughts I wanted to throw out there! My ADHD brain tends to wander toward questions like 'How do we improve 133?' and 'Why?' in the background. I am curious to hear what you think! Either way, looking forward to class.